include/circt/Dialect/HW/HWAttributesNaming.td

```diff
@@ -17,32 +17,65 @@ include "circt/Dialect/HW/HWDialect.td"
 include "mlir/IR/AttrTypeBase.td"
 
 def InnerRefAttr : AttrDef<HWDialect, "InnerRef"> {
-  let summary = "Refer to a name inside a module";
+  let summary = "Refer to a name inside a list of scopes";
   let description = [{
-    This works like a symbol reference, but to a name inside a module.
+    This works like a symbol reference, but to a name inside a list of `InnerSymbolTable`'s.
   }];
   let mnemonic = "innerNameRef";
-  let parameters = (ins "::mlir::FlatSymbolRefAttr":$moduleRef,
-                        "::mlir::StringAttr":$name);
+  let parameters = (ins ArrayRefParameter<"::mlir::StringAttr">:$path);
 
   let builders = [
-    AttrBuilderWithInferredContext<(ins "::mlir::StringAttr":$module,
-                                        "::mlir::StringAttr":$name), [{
-      return $_get(
-          module.getContext(), mlir::FlatSymbolRefAttr::get(module), name);
-    }]>
+
+    AttrBuilderWithInferredContext<(ins "::mlir::ArrayRef<::mlir::StringAttr>":$pathAttr), [{
+      assert(!pathAttr.empty() && "Empty path");
+      assert(pathAttr.front() && "Null path element");
+      return InnerRefAttr::get(pathAttr.front().getContext(), pathAttr);
+    }]>,
+  
+    AttrBuilderWithInferredContext<(ins "::mlir::ArrayRef<::mlir::StringAttr>":$pathAttr,
+                                        "::mlir::StringAttr":$target), [{
+      llvm::SmallVector<::mlir::StringAttr> fullPath(pathAttr.begin(), pathAttr.end());
+      fullPath.push_back(target);
+      return InnerRefAttr::get(target.getContext(), fullPath);
+    }]>,
+
+    // @mortbopet: Old InnerRef single-module compatability builders.
+    AttrBuilderWithInferredContext<(ins "::mlir::StringAttr":$target), [{
+      llvm::SmallVector<::mlir::StringAttr> fullPath;
+      fullPath.push_back(target);
+      return InnerRefAttr::get(fullPath);
+    }]>,
+
+    AttrBuilderWithInferredContext<(ins 
+                                    "::mlir::StringAttr":$parent,
+                                    "::mlir::StringAttr":$target), [{
+      ::llvm::SmallVector<::mlir::StringAttr> fullPath{parent};
+      fullPath.push_back(target);
+      return InnerRefAttr::get(fullPath);
+    }]>,
+    AttrBuilder<(ins "::mlir::StringAttr":$parent,
+                     "::mlir::StringAttr":$target), [{
+      return InnerRefAttr::get(parent, target);
+    }]>,
   ];
 
   let hasCustomAssemblyFormat = 1;
+  let genVerifyDecl = 1;
 
   let extraClassDeclaration = [{
-    /// Get the InnerRefAttr for an operation and add the sym on it.
-    static InnerRefAttr getFromOperation(mlir::Operation *op,
-                                         mlir::StringAttr symName,
-                                         mlir::StringAttr moduleName);
+    ::mlir::StringAttr getRoot() const {
+      assert(!getPath().empty() && "Empty path");
+      return getPath().front();
+    }
+
+    ::mlir::FlatSymbolRefAttr getRootRef() const {
```

Comment:
Although nothing complains, per-se, FlatSymbolRefAttr should only be used to refer to a real symbol, we shouldn't just shove any old string in one.  A string array as a path with this explicit semi-cast is fine, if the verifier checks that the first element of the path resolves to a symbol.

Reply:
Will remove and change at current `getRootRef` usage sites - given that the first element of the path isn't guaranteed to resolve to a symbol.